File: comparisons/lmm_penicillin.py
"""
Penicillin crossed random effects using statsmodels.

`lme4::lmer(diameter ~ 1 + (1 | plate) + (1 | sample))` has two crossed random
intercepts. `MixedLM` allows one grouping factor per fit; we fit the plate
stratum only as a runnable benchmark (not identical to the full crossed model).
"""

from __future__ import annotations

import os
import sys

try:
    import pandas as pd
    import statsmodels.formula.api as smf
except ImportError:
    print("Please install pandas and statsmodels:")
    print("pip install pandas statsmodels")
    sys.exit(1)


def main() -> None:
    path = os.path.join("tests", "data", "penicillin.csv")
    if not os.path.exists(path):
        print(f"Could not find the dataset at {path}")
        print("Please run this example from the root of the lme-rs repository.")
        sys.exit(1)

    data = pd.read_csv(path)
    print("\nFitting model: diameter ~ 1 + (1 | plate)  [single grouping; crossed RE not in one MixedLM]")
    model = smf.mixedlm("diameter ~ 1", data, groups=data["plate"])
    result = model.fit(reml=True)
    print("\n=== Model Summary ===")
    print(result.summary())

    print("\n=== Predictions (population-level) ===")
    fe = result.fe_params
    print("Intercept:", float(fe["Intercept"]))


if __name__ == "__main__":
    main()

File: comparisons/lmm_penicillin.R
library(lme4)

# 1. Load the dataset
data <- read.csv("tests/data/penicillin.csv")
data$plate <- as.factor(data$plate)
data$sample <- as.factor(data$sample)

# 2. Fit the model
cat("\nFitting model: diameter ~ 1 + (1 | plate) + (1 | sample)\n")
fit <- lmer(diameter ~ 1 + (1 | plate) + (1 | sample), data = data, REML = TRUE)

# 3. Print the summary
cat("\n=== Model Summary ===\n")
print(summary(fit))

# 4. Generate Predictions
cat("\n=== Predictions ===\n")
cat("Generating predictions for new plates and samples...\n")

newdata <- data.frame(
  plate = factor(c("a", "b", "c", "d"), levels = levels(data$plate)),
  sample = factor(c("A", "C", "E", "F"), levels = levels(data$sample))
)

# Predict population-level (re.form = NA)
preds <- predict(fit, newdata, re.form = NA)
cat("Predictions:\n")
print(preds)
